Add Neon blog post (#11)
IAmJSD/website: src/pages/blog/neon.md

---
layout: ../../layouts/BlogPostLayout.astro

> title: "Setting up a multi-tenant database project on Rails using Neon"

Reviewer: ```suggestion
title: "Setting up a database-per-tenant project on Rails using Neon"
```

> description: "How to setup a multi-tenant project within Ruby on Rails."

Reviewer: ```suggestion
description: "How to set up a Rails project where each user's data is stored in a separate Postgres database on Neon."
```

---

**Legal disclosure:** My company (Web Scale Software Ltd) was paid by Neon to write this blog post. This is an advert.

> In this tutorial, we are going to build a basic notes application using a multi-tenant database architecture with Ruby on Rails. This means that each user has their own database that can individually scale.

Reviewer: ```suggestion
In this tutorial, we are going to build a basic notes application using a database-per-user database architecture with Ruby on Rails. This means that each user has their own database that can individually scale.
```

## Installation

Firstly, make sure you have at least Ruby 3.1 installed:
```
➜  ~ ruby --version
ruby 3.2.3 (2024-01-18 revision 52bb2ac0a6) [x86_64-linux-gnu]
```

If you do not have a conpatible version, you may need to install rbenv and then use that to get the latest Ruby release. `rbenv` is suggested to stop any weird behaviours that a system package manager may cause.

At this point, install the `rails` gem using `gem install rails`.
From here, we can make a new Rails project with `rails new --database=postgresql note-taking`. We will make it using Postgres as the database so that we can setup Neon easily:

```
➜  ~ rails new --database=postgresql note-taking
      create
...
```

> After this is done, we should have a Hotwired application running on the latest release of Rails. Note to use Postgres, you may need to install your OS's dependency for it (`libpg-dev` in the case of Linux).

Reviewer: ```suggestion
After this is done, we should have a Hotwired application running on the latest release of Rails. Note: to use Postgres, you may need to install your OS's dependency for it (`libpg-dev` in the case of Linux).
```

- **Data Compliance:** If in the future we wanted to deploy in a region that the user specifies for compliance reasons (for example, an American user might want to deploy in America if you have a point of presence there due to latency reasons, but a European user might want to deploy in Europe due to Neon's GDPR compliance), this is a drop down box away with this architecture. Additionally, data compliance is significantly easier here. Want to add GDPR exports? Just dump out the entire database. Want to delete a user without worrying about errors due to foreign keys within your large database? Just delete the database (and we will tie this to the User record in this tutorial).
- **Competitive Advantage:** With the multi-database model, you don't have to worry about leaking information that whilst it may not be sensitive to your users, is quite revealing about your company. For example, in this case, you might be able to figure out the companies user retention by taking a publicly sent customer count and then comparing it to the note ID since it is auto-incrementing with ActiveRecord.

> The main tradeoff with this model is that migrations become more complex, although this can be managed with careful branching and scripting.

Reviewer: ```suggestion
The main tradeoff with this model is that migrations become more complex, although we'll show how this can be managed with careful branching and scripting.
```

To do this, we are going to need to make a few network requests to the Neon API. Since `net/http` is capped to HTTP/1.1, we are going to use `httpx` to do this faster. Add this to your `Gemfile` and then run `bundle`:

```rb
# Add support for higher than HTTP/1.1 clients.
gem 'httpx', '~> 1.3.1'
```

> To do this, we want to create a database and tie it to this record until it is destroyed. In case we later want to make a database per another entity (for example, one per organisation), we will make this a concern that we can share across anything that needs it. To do this, we will remove `app/models/concerns/.keep` and create a `multi_tenant.rb` file there to manage this:

Reviewer: ```suggestion
We want to create a database and tie it to this record until it is destroyed. In case we later want to make a database per another entity (for example, one per organisation), we will make this a concern that we can share across anything that needs it. To do this, we will remove `app/models/concerns/.keep` and create a `multi_tenant.rb` file there to manage this:
```

    # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
    devise :database_authenticatable, :registerable,
        :recoverable, :rememberable, :validatable
end
```

Now we should go over to the branch we want to base off (in this case, `main`, but in staging this might be `develop` or something else) and select it on the dashboard. We should then add a role:

![The roles and database page](/images/neon/role_database_page.png)

![Role create page](/images/neon/role_create.png)

When you get to the success screen, you should add this into the `.env.development` file:

```
# Defines the role for the multi-tenant architecture.
NEON_ROLE_NAME=<role name>
NEON_ROLE_PASSWORD=<role password>
```

Now we should make the user database and set the owner to this role:

![DB create page](/images/neon/db_create.png)

Now we should set the database name and other Neon information (you can find the project ID in settings and API tokens in account settings):

```
# Set the user database name.
USER_DATABASE_NAME=user_db

# Defines the Neon API token and project ID.
NEON_API_TOKEN=<api token here>
NEON_PROJECT_ID=<project ID here>
```

> Great! Now we have a User that repersents the lifetime of the database tenant. When the user is destroyed, the database will be too.

Reviewer: ```suggestion
Great! Now we have a User that represents the lifetime of the database tenant. When the user is destroyed, the database will be too.
```

## Building the Note record and handling migrations

To build the Note model, we will firstly just make the model in a traditional way:

```
➜  note-taking git:(main) rails g model note title:text body:text
      invoke  active_record
      create    db/migrate/20241013152402_create_notes.rb
      create    app/models/note.rb
      invoke    test_unit
      create      test/models/note_test.rb
      create      test/fixtures/notes.yml
```

> However, before we migrate the records, we need to handle multi-tenancy migrations. To do this, we will use [neon-bulk-migrator](https://github.com/IAmJSD/neon-bulk-migrator). To do so, we need to add the following to `database.yml`:

Reviewer: ```suggestion
However, before we migrate the records, we need to handle migrations. To do this, we will use [neon-bulk-migrator](https://github.com/IAmJSD/neon-bulk-migrator). To do so, we need to add the following to `database.yml`:
```

```rb
class ApplicationController < ActionController::Base
  ...

  # Ensure the user is authenticated
  before_action :authenticate_user!
end
```

Now we will make the `NotesController`:

```
➜  note-taking git:(main) rails g controller notes
      create  app/controllers/notes_controller.rb
      invoke  erb
      create    app/views/notes
      invoke  test_unit
      create    test/controllers/notes_controller_test.rb
      invoke  helper
      create    app/helpers/notes_helper.rb
      invoke    test_unit
```

From here, we will add a new index by adding this to `config/routes.rb`:

```rb
# Make the root the main application page.
root "notes#all"
```

and just a simple hello world to `app/views/notes/all.html.erb`:

```html
<h1>Hello World!</h1>
```

From here, when you go to `/`, we will see a user login screen. Lets sign up and then we will get our first user. If we open `rails c`, we will be able to see this:

```
➜  note-taking git:(main) rails c
Loading development environment (Rails 7.2.1)
note-taking(dev)> User.first!
  User Load (204.5ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT $1  [["LIMIT", 1]]
=> #<User id: 1, ...
```

> Now we are going to build the controller for this view. Note that in production you likely would want to add pagination to the accessing of notes to improve page load time. We want to access within the `database` block since this will handle the multi-tenant behaviour:

Reviewer: ```suggestion
Now we are going to build the controller for this view. Note that in production you likely would want to add pagination to the accessing of notes to improve page load time. We want to access within the `database` block since this will handle the db-per-tenant behaviour:
```

    </div>
    <%= form.submit %>
<% end %>
```

We can now add this to `config/routes.rb`:

```rb
# Add the new form handler.
get "notes/new", to: "notes#new"
post "notes/new", to: "notes#create"
get "notes/:id", to: "notes#show"
```

And add the `create` method to the controller:

```rb
    def create
        @note = current_user.database do
            Note.create(params.permit(:title, :body))
        end
        return render :new, status: :bad_request if @note.errors.any?
        redirect_to "/notes/#{@note.id}"
    end
```

And we will want to add constraints to our Note model:

```rb
class Note < ApplicationRecord
    validates :title, presence: true, length: { minimum: 2 }
    validates :body, presence: true, length: { minimum: 2 }
end
```

Next, we'll add the `show` method to the controller:

```rb
    def show
        @note = current_user.database do
            Note.find params[:id]
        end
    end
```

And then make `app/views/notes/show.html.erb`:

```html
<h1><%= @note.title %></h1>

<p><pre><%= @note.body %></pre></p>

<p>
    <a href="/">Show all notes</a>
</p>
```

You now have a basic multi-tenant notes application using Ruby on Rails!

## Conclusion

> In conclusion, whilst this method of database storage still has its compromises which you need to consider, this shows a new way of storing data that offers different data protection, data integrity, and potentially cost benefits since each user can be individually scaled up and down without needing to worry about noisy neighbours.

Reviewer: ```suggestion
In conclusion, whilst this method of database storage still has its tradeoffs which you need to consider, this shows a new way of storing data that offers different data protection, data integrity, and potentially cost benefits since each user can be individually scaled up and down without needing to worry about noisy neighbours.
```

---
layout: ../../layouts/BlogPostLayout.astro
title: "Setting up a multi-tenant database project on Rails using Neon"
description: "How to setup a multi-tenant project within Ruby on Rails."
date: "16 Oct 2024 17:00 GMT"
---

**Legal disclosure:** My company (Web Scale Software Ltd) was paid by Neon to write this blog post. This is an advert.

> In this tutorial, we are going to build a very notes application using a multi-tenant database architecture with Ruby on Rails. This means that each user has their own database that can individually scale.

Reviewer: ```suggestion
In this tutorial, we are going to build a basic notes application using a multi-tenant database architecture with Ruby on Rails. This means that each user has their own database that can individually scale.
```

      create    db/migrate/20241012124645_create_users.rb
      create    app/models/user.rb
      invoke    test_unit
      create      test/models/user_test.rb
      create      test/fixtures/users.yml
➜  note-taking git:(main) rails g devise User
      invoke  active_record
      create    db/migrate/20241012124742_add_devise_to_users.rb
      insert    app/models/user.rb
       route  devise_for :users
```

Next, we will run `rails db:migrate` to store our changes and add `database_information` to `config/initializers/filter_parameter_logging.rb`:
```rb
# Be sure to restart your server when you modify this file.

# Configure parameters to be partially matched (e.g. passw matches password) and filtered from the log file.
# Use this to limit dissemination of sensitive information.
# See the ActiveSupport::ParameterFilter documentation for supported notations and behaviors.
Rails.application.config.filter_parameters += [
  ...,
  :database_information,
]
```

So now we have our users, but what about our notes? Let's add logic to put each user's notes in a separate Postgres database.

There are several advantages to  this database-per-tenant architecture:

> - **Security by Design:** There is significantly less risk of harm due to accidentally destroying/fetching other customers data. This is because instead of worrying about multiple columns when you do things to a users notes, you just have to think about one. Additionally, you don't have to worry about the risk of someone finding another users notes by decrementing their own note ID since the users notes are pulled from the signed in record.

Reviewer: ```suggestion
- **Security by Design:** There is less risk of harm due to accidentally destroying/fetching other customer's data. This is because instead of worrying about multiple columns when you do things to a user's notes, you just have to think about one. Additionally, you have less risk of someone finding another users notes _( for example, by decrementing their own note ID)_ since the user's notes are pulled from the signed in record.
```
